### PR TITLE
RHMAP-6712 add export read and status commands

### PR DESCRIPTION
**Pull request by pb82**
# Motivation

Create `read` and `status` commands for appdata export in fhc. `read` will just return the job definition as JSON. `status` will do a realtime polling of the job status until it's finished or the user aborts. The syntax is:

`fhc appdata export read`
`fhc appdata export status`

`status` will print and update a progress bar. I decided to use the not-so-fance `progress` packate instead of the more-fance `node-progress-bars` because the former one does not have any additional dependencies.

ping @odra @paolobueno @wei-lee

### CODE REVIEW

File: lib/cmd/fh3/appdata/export/status.js [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+/**
+ * Update the progressbar if possible. Don't use `bar#tick` as that will not
+ * set the current progress but add the new valud to the current progress.
+ */
+function printout(data, bar) {
+  if (data.totalSteps && data.step >= 0) {
+    bar.update(data.step / data.totalSteps);
+  } else {
+    /*ignore jslint start*/
+    console.log('No progress information in response');
```
Comment (fh-qube) [inline lib/cmd/fh3/appdata/export/status.js]: MAJOR Remove this logging statement. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS2228)

File: lib/cmd/fh3/appdata/export/status.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+      readProgress(params, function (err, result) {
+        if (err) {
+          clearInterval(poller);
+          return cb(err);
+        }
+
+        // supercore sends the actual result in the `data` field
+        result = result.data;
+
+        if (bar == null && result.totalSteps) {
```
Comment (fh-qube) [inline lib/cmd/fh3/appdata/export/status.js]: MAJOR Move the declaration of "bar" before this usage. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AVariableDeclarationAfterUsage)

File: lib/cmd/fh3/appdata/export/status.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    /*ignore jslint start*/
+    // Need to use `console#log` here until we introduce some other logging or outpur
+    // framework to fhc
+    console.log('Polling job progress. Press `q` to quit');
+    /*ignore jslint end*/
+
+    function onInterval() {
+      readProgress(params, function (err, result) {
+        if (err) {
+          clearInterval(poller);
```
Comment (fh-qube) [inline lib/cmd/fh3/appdata/export/status.js]: MAJOR Move the declaration of "poller" before this usage. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AVariableDeclarationAfterUsage)

File: lib/cmd/fh3/appdata/export/status.js [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    'appId': 'Application id',
+    'jobId': 'Export job id',
+    'interval': 'Polling interval in milliseconds (default is 2000)'
+  },
+  'method': 'get',
+  'customCmd': function (params, cb) {
+    /*ignore jslint start*/
+    // Need to use `console#log` here until we introduce some other logging or outpur
+    // framework to fhc
+    console.log('Polling job progress. Press `q` to quit');
```
Comment (fh-qube) [inline lib/cmd/fh3/appdata/export/status.js]: MAJOR Remove this logging statement. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS2228)

**Comment by fh-qube** [comment]
SonarQube analysis reported 4 issues:
* MAJOR 4 major

Watch the comments in this conversation to review them.

**Comment by fh-automatron** [comment]
### Pull Request Quality Guidelines

As we continue to focus on quality, here are a few things to keep in mind:

* The code changes conform to the contributing & code style guides for this component (if they exist)
* A component owner should review your change (we've tagged these above)
* Automated testing has been added or updated for this change (e.g. UART, Unit/Acceptance tests)
* There should be an associated fhcap PR with this change - please provide a link to this
* If updating server component this PR should be deployed to a grid somewhere (via `fhcap-cli`) - please provide a link to this
* Ensure JIRA is up to date before & after merge (e.g. deployment notes, tester notes, release notes, fix version)
* Any SonarQube issues should be addressed where applicable (use your own judgement)

Please ping a component owner (see `fh contrib`) when your builds are complete, changes are deployed and you're ready for a review.

In order to update fhcap with your component change please execute:
`fhcap component update fh-fhc 2.7.6 324`

File: lib/cmd/fh3/appdata/export/status.js [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    /*ignore jslint end*/
+
+    var bar = null;
+    function onInterval() {
+      readProgress(params, function (err, result) {
+        if (err) {
+          clearInterval(poller);
```
Comment (fh-qube) [inline lib/cmd/fh3/appdata/export/status.js]: MAJOR Move the declaration of "poller" before this usage. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AVariableDeclarationAfterUsage)

**Comment by fh-qube** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 3 major

Watch the comments in this conversation to review them.

**Comment by fh-qube** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

**Comment by wei-lee** [comment]
LGTM

File: lib/cmd/fh3/appdata/export/status.js [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+          return cb(err);
+        }
+
+        // supercore sends the actual result in the `data` field
+        result = result.data;
+
+        if (bar == null && result.totalSteps) {
```
Comment (vchepeli) [inline lib/cmd/fh3/appdata/export/status.js]: this should be in new function

File: lib/cmd/fh3/appdata/export/start.js [hunk 7]
```diff
@@ -44,7 +44,15 @@ module.exports = {
         return cb(err);
       }
 
-      return cb(null, 'Created a new export job with ID: ' + result);
+      var info = [
+        'Created a new export job with ID: ' + result,
+        '. You can use',
+        ' `fhc appdata export status --envId=' + params.envId,
+        ' --appId=' + params.appId + ' --jobId=' + result,
+        '` to poll the job status'
```
Comment (paolobueno) [inline lib/cmd/fh3/appdata/export/start.js]: I suggest removing the backticks and trying to get the command on a separate line so it's easier for the user to copy:

```
Created a new ... To check on the job status you can use the following command:
fhc appdata export ...
```

File: lib/cmd/fh3/appdata/export/status.js [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+    // Need to use `console#log` here until we introduce some other logging or outpur
+    // framework to fhc
+    console.log('Polling job progress. Press `q` to quit');
+    /*ignore jslint end*/
+
+    var bar = null, poller = null;
+
+    function onInterval() {
+      readProgress(params, function (err, result) {
+        if (err) {
```
Comment (vchepeli) [inline lib/cmd/fh3/appdata/export/status.js]: duplicate code here and same on line 80. create new function

File: lib/cmd/fh3/appdata/export/status.js [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+    // Raw mode is required to flush stdin without
+    // pressing return
+    process.stdin.setRawMode(true);
+    process.stdin.on('readable', function () {
+      var chunk = process.stdin.read();
+
+      // `stdin#read` can return null
+      if (chunk) {
+        var letter = chunk.toString();
+        if (EXIT_KEYS.indexOf(letter) >= 0) {
```
Comment (vchepeli) [inline lib/cmd/fh3/appdata/export/status.js]: same code here. You have condition and {clearInterval(poller);return cb(null);}

File: lib/cmd/fh3/appdata/export/start.js (hunk 7 above)
Comment (vchepeli) [inline lib/cmd/fh3/appdata/export/start.js]: Yep, that would be much better

File: lib/cmd/fh3/appdata/export/status.js (hunk 4 above)
Comment (paolobueno) [inline lib/cmd/fh3/appdata/export/status.js]: maybe also add `/* eslint disable-next-line no-console */`: [eslint docs](http://eslint.org/docs/user-guide/configuring#disabling-rules-with-inline-comments)

File: npm-shrinkwrap.json [hunk 10]
```diff
@@ -1,6 +1,6 @@
 {
```
Comment (vchepeli) [inline npm-shrinkwrap.json]: Ahh. Bumping versions should be in different commit 👎

File: lib/cmd/fh3/appdata/export/status.js (hunk 4 above)
Comment (vchepeli) [inline lib/cmd/fh3/appdata/export/status.js]: you should try it if it works for this case

**Comment by vchepeli** [comment]
LGTM

**Comment by paolobueno** [comment]
LGTM :+1:

File: lib/cmd/fh3/appdata/export/start.js (hunk 7 above)
Comment (pb82) [inline lib/cmd/fh3/appdata/export/start.js]: Alright, going to change that

File: lib/cmd/fh3/appdata/export/status.js [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+/**
+ * Update the progressbar if possible. Don't use `bar#tick` as that will not
+ * set the current progress but add the new valud to the current progress.
+ */
+function printout(data, bar) {
+  if (bar && data.totalSteps && data.step >= 0) {
+    bar.update(data.step / data.totalSteps);
+  } else {
+    /* eslint disable-next-line no-console */
+    console.log('No progress information in response');
```
Comment (fh-qube) [inline lib/cmd/fh3/appdata/export/status.js]: MAJOR Remove this logging statement. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS2228)

File: lib/cmd/fh3/appdata/export/status.js [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+      if (poller) {
+        clearInterval(poller);
+      }
+      return cb(null);
+    }
+
+    // Need to use `console#log` here until we introduce some other logging or outpur
+    // framework to fhc
+    /* eslint disable-next-line no-console */
+    console.log('Polling job progress. Press `q` to quit');
```
Comment (fh-qube) [inline lib/cmd/fh3/appdata/export/status.js]: MAJOR Remove this logging statement. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS2228)

**Comment by fh-qube** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

File: lib/cmd/fh3/appdata/export/status.js (hunk 9 above)
Comment (pb82) [inline lib/cmd/fh3/appdata/export/status.js]: Adressed that with the `cleanupAndExit` function
